hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/states/endpoint/UnRegisterEndpointTask.java

```diff
@@ -0,0 +1,262 @@
+    Callable<EndpointStateMachine.EndPointStates> {
+  static final Logger LOG =
+          LoggerFactory.getLogger(UnRegisterEndpointTask.class);
+
+  private final EndpointStateMachine rpcEndPoint;
+  private final ConfigurationSource conf;
+  private Future<EndpointStateMachine.EndPointStates> result;
+  private DatanodeDetails datanodeDetails;
+  private final OzoneContainer datanodeContainerManager;
+  private StateContext stateContext;
+
+  /**
+   * Creates a register endpoint task.
+   *
+   * @param rpcEndPoint - endpoint
+   * @param conf - conf
+   * @param ozoneContainer - container
+   */
+  @VisibleForTesting
+  public UnRegisterEndpointTask(EndpointStateMachine rpcEndPoint,
+      ConfigurationSource conf, OzoneContainer ozoneContainer,
+      StateContext context) {
+    this.rpcEndPoint = rpcEndPoint;
+    this.conf = conf;
+    this.datanodeContainerManager = ozoneContainer;
+    this.stateContext = context;
+
+  }
+
+  /**
+   * Get the DatanodeDetails.
+   *
+   * @return DatanodeDetailsProto
+   */
+  public DatanodeDetails getDatanodeDetails() {
+    return datanodeDetails;
+  }
+
+  /**
+   * Set the contiainerNodeID Proto.
+   *
+   * @param datanodeDetails - Container Node ID.
+   */
+  public void setDatanodeDetails(
+      DatanodeDetails datanodeDetails) {
+    this.datanodeDetails = datanodeDetails;
+  }
+
+  /**
+   * Computes a result, or throws an exception if unable to do so.
+   *
+   * @return computed result
+   * @throws Exception if unable to compute a result
+   */
+  @Override
+  public EndpointStateMachine.EndPointStates call() throws Exception {
+
+    if (getDatanodeDetails() == null) {
+      LOG.error("DatanodeDetails cannot be null in RegisterEndpoint task, "
+          + "shutting down the endpoint.");
+      return rpcEndPoint.setState(EndpointStateMachine.EndPointStates.SHUTDOWN);
+    }
+
+    rpcEndPoint.lock();
+    try {
+
+      if (rpcEndPoint.getState()
+          .equals(EndPointStates.SHUTDOWN)) {
+        ContainerReportsProto containerReport =
+            datanodeContainerManager.getController().getContainerReport();
+        NodeReportProto nodeReport = datanodeContainerManager.getNodeReport();
+        PipelineReportsProto pipelineReportsProto =
+            datanodeContainerManager.getPipelineReport();
+        // TODO : Add responses to the command Queue.
+        SCMUNRegisteredResponseProto response = rpcEndPoint.getEndPoint()
+            .unregister(datanodeDetails.getProtoBufMessage(), nodeReport,
+                containerReport, pipelineReportsProto);
+        Preconditions.checkState(UUID.fromString(response.getDatanodeUUID())
+                .equals(datanodeDetails.getUuid()),
+            "Unexpected datanode ID in the response.");
+        Preconditions.checkState(!StringUtils.isBlank(response.getClusterID()),
+            "Invalid cluster ID in the response.");
+        if (response.hasHostname() && response.hasIpAddress()) {
```

Comment:
Not sure if the datanodeDetails updating is necessary for this command is a unregister command which is a last connection to SCM.

Reply:
done